### PR TITLE
Module imagetools

### PR DESCRIPTION
**Pull request by n2pro**

### CODE REVIEW

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/ThumbnailGeneration/ImageChange.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.ImageToolsModule.Core.ThumbnailGeneration
+{
+    public class ImageChange
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/ThumbnailGeneration/ImageChange.cs]: Need to inherit from ValueObject

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/ThumbnailGeneration/ThumbnailGenerationResult.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace VirtoCommerce.ImageToolsModule.Core.ThumbnailGeneration
+{
+    public class ThumbnailGenerationResult
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/ThumbnailGeneration/ThumbnailGenerationResult.cs]: Inherit from valueObject

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/ThumbnailGeneration/ThumbnailTaskProgress.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace VirtoCommerce.ImageToolsModule.Core.ThumbnailGeneration
+{
+    public class ThumbnailTaskProgress
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/ThumbnailGeneration/ThumbnailTaskProgress.cs]: Inherit from ValueObject

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using VirtoCommerce.ImageToolsModule.Core.Models;
+using VirtoCommerce.ImageToolsModule.Core.Services;
+using VirtoCommerce.ImageToolsModule.Data.Repositories;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.ImageToolsModule.Data.Services
+{
+    public class ThumbnailOptionSearchService : ServiceBase, IThumbnailOptionSearchService
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs]: Inheritance from ServiceBase no longer needed

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+        public ThumbnailOptionSearchService(Func<IThumbnailRepository> thumbnailRepositoryFactory)
+        {
+            this._thumbnailRepositoryFactory = thumbnailRepositoryFactory;
+        }
+
+        public async Task<ThumbnailOptionSearchResult> SearchAsync(ThumbnailOptionSearchCriteria criteria)
+        {
+            using (var repository = this._thumbnailRepositoryFactory())
+            {
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs]: add repositorty.DisableCahngesTracking()

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+using System.Threading.Tasks;
+using VirtoCommerce.ImageToolsModule.Core.Models;
+using VirtoCommerce.ImageToolsModule.Core.Services;
+using VirtoCommerce.ImageToolsModule.Data.Models;
+using VirtoCommerce.ImageToolsModule.Data.Repositories;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.ImageToolsModule.Data.Services
+{
+    public class ThumbnailOptionService : ServiceBase, IThumbnailOptionService
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs]: Inheritance from ServiceBase no longer needed

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+                repository.UnitOfWork.Commit();
+                pkMap.ResolvePrimaryKeys();
+            }
+        }
+
+        public async Task<ThumbnailOption[]> GetByIdsAsync(string[] ids)
+        {
+            using (var repository = this._thumbnailRepositoryFactory())
+            {
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs]: Add repository.DisableChangesTracking()

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskSearchService.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                    sortInfos = new[]
+                    {
+                        new SortInfo
+                        {
+                            SortColumn = ReflectionUtility.GetPropertyName<ThumbnailTask>(t => t.CreatedDate), SortDirection = SortDirection.Descending
+                        }
+                    };
+                }
+                query = query.OrderBySortInfos(sortInfos);
+                var totalCount = query.Count();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskSearchService.cs]: CountAsync()

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskSearchService.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+                var ids = query.Skip(criteria.Skip).Take(criteria.Take).Select(x => x.Id).ToArray();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskSearchService.cs]: ToArrayAsync()

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            var snapshot = serviceCollection.BuildServiceProvider();
+            var configuration = snapshot.GetService<IConfiguration>();
+
+            serviceCollection.AddDbContext<ThumbnailDbContext>(options => options.UseSqlServer(configuration.GetConnectionString("VirtoCommerce")));
+
+            serviceCollection.AddScoped<IThumbnailRepository, ThumbnailRepository>();
+            serviceCollection.AddScoped<Func<IThumbnailRepository>>(provider => () => provider.CreateScope().ServiceProvider.GetService<IThumbnailRepository>());
+
+
+            serviceCollection.AddTransient<IThumbnailOptionService, ThumbnailOptionService>();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs]: All these services should be registered as singleton

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        public ManifestModuleInfo ModuleInfo { get; set; }
+        public void Initialize(IServiceCollection serviceCollection)
+        {
+            var snapshot = serviceCollection.BuildServiceProvider();
+            var configuration = snapshot.GetService<IConfiguration>();
+
+            serviceCollection.AddDbContext<ThumbnailDbContext>(options => options.UseSqlServer(configuration.GetConnectionString("VirtoCommerce")));
+
+            serviceCollection.AddScoped<IThumbnailRepository, ThumbnailRepository>();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs]: Transient

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+            serviceCollection.AddScoped<Func<IThumbnailRepository>>(provider => () => provider.CreateScope().ServiceProvider.GetService<IThumbnailRepository>());
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs]: Singleton

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+
+            serviceCollection.AddTransient<IThumbnailOptionService, ThumbnailOptionService>();
+            serviceCollection.AddTransient<IThumbnailOptionSearchService, ThumbnailOptionSearchService>();
+
+            serviceCollection.AddTransient<IThumbnailTaskSearchService, ThumbnailTaskSearchService>();
+            serviceCollection.AddTransient<IThumbnailTaskService, ThumbnailTaskService>();
+            serviceCollection.AddTransient<IPushNotificationManager, PushNotificationManager>();
+            serviceCollection.AddTransient<IUserNameResolver, HttpContextUserResolver>();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs]: UserNameRoslver mustn't be here

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        public void PostInitialize(IServiceProvider serviceProvider)
+        {
+            AbstractTypeFactory<ThumbnailOption>.RegisterType<ThumbnailOption>().MapToType<ThumbnailOptionEntity>();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Module.cs]: These lines are redundnat

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/ThumbnailConstants.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using VirtoCommerce.Platform.Core.Modularity;
+
+namespace VirtoCommerce.ImageToolsModule.Core
+{
+    public static class ThumbnailConstants
+    {
+        public static class Permissions
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/ThumbnailConstants.cs]: Wrap Permissions static class by Security

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+                var thumbnailOptions = await repository.GetThumbnailOptionsByIdsAsync(ids);
+                return thumbnailOptions.Select(x => x.ToModel(AbstractTypeFactory<ThumbnailOption>.TryCreateInstance())).ToArray();
+            }
+        }
+
+        public Task RemoveByIdsAsync(string[] ids)
+        {
+            using (var repository = this._thumbnailRepositoryFactory())
+            {
+                repository.RemoveThumbnailOptionsByIds(ids);
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs]: Calling async methods without await leads to dispose DbContext and exceptions inside repository

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/Models/ThumbnailOptionSearchResult.cs [hunk 17]
```diff
@@ -0,0 +1,11 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.ImageToolsModule.Core.Models
+{
+    public class ThumbnailOptionSearchResult : GenericSearchResult<ThumbnailOption>
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/Models/ThumbnailOptionSearchResult.cs]: Can we get rid of this type? Use GenericSearcResult instead?

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/Models/ThumbnailTaskSearchResult.cs [hunk 18]
```diff
@@ -0,0 +1,12 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.ImageToolsModule.Core.Models
+{
+    public class ThumbnailTaskSearchResult : GenericSearchResult<ThumbnailTask>
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/Models/ThumbnailTaskSearchResult.cs]: Can we get rid of this type? Use GenericSearcResult instead?

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+                    sortInfos = new[]
+                    {
+                        new SortInfo
+                        {
+                            SortColumn = ReflectionUtility.GetPropertyName<ThumbnailTask>(t => t.CreatedDate), SortDirection = SortDirection.Descending
+                        }
+                    };
+
+                var query = repository.ThumbnailOptions.OrderBySortInfos(sortInfos);
+                var totalCount = query.Count();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs]: CountAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
+                var ids = query.Skip(criteria.Skip).Take(criteria.Take).Select(x => x.Id).ToArray();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs]: ToArrayAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+                            sourceOptionsEntity.Patch(targetOptionsEntity);
+                        }
+                        else
+                        {
+                            repository.Add(sourceOptionsEntity);
+                        }
+                    }
+                }
+
+                repository.UnitOfWork.Commit();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs]: ComitASync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+                return thumbnailOptions.Select(x => x.ToModel(AbstractTypeFactory<ThumbnailOption>.TryCreateInstance())).ToArray();
+            }
+        }
+
+        public async Task RemoveByIdsAsync(string[] ids)
+        {
+            using (var repository = this._thumbnailRepositoryFactory())
+            {
+                await repository.RemoveThumbnailOptionsByIds(ids);
+                repository.UnitOfWork.Commit();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionService.cs]: ComitAync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskService.cs [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+                            sourceTaskEntity.Patch(targetTaskEntity);
+                        }
+                        else
+                        {
+                            repository.Add(sourceTaskEntity);
+                        }
+                    }
+                }
+
+                repository.UnitOfWork.Commit();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskService.cs]: COmitAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/ThumbnailGeneration/ThumbnailGenerationProcessor.cs [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+                foreach (var task in tasks)
+                {
+                    var changedSince = regenerate ? null : task.LastRun;
+                    progressInfo.TotalCount = await _imageChangesProvider.GetTotalChangesCount(task, changedSince, token);
+                }
+            }
+
+            progressCallback(progressInfo);
+
+            var pageSize = _settingsManager.GetValue("ImageTools.Thumbnails.ProcessBacthSize", 50);
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/ThumbnailGeneration/ThumbnailGenerationProcessor.cs]: We must use strongly typed settings everywhere

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Localizations/en.VirtoCommerce.Thumbnail.json [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+            "setting-save": {
+                "title": "Save changes",
+                "message": "The Setting has been modified. Do you want to save changes?"
+            },
+            "setting-delete": {
+                "title": "Delete confirmation",
+                "message": "Are you sure you want to delete this Setting?"
+            }
+        },
+        "commands": {
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Web/Localizations/en.VirtoCommerce.Thumbnail.json]: Missed localization for Settings and Permissions

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/Services/IThumbnailTaskService.cs [hunk 26]
```diff
@@ -0,0 +1,15 @@
...
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using VirtoCommerce.ImageToolsModule.Core.Models;
+
+namespace VirtoCommerce.ImageToolsModule.Core.Services
+{
+    public interface IThumbnailTaskService
+    {
+        Task SaveOrUpdateAsync(ICollection<ThumbnailTask> options);
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Core/Services/IThumbnailTaskService.cs]: Rename to SaveChangesAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Repositories/ThumbnailRepository.cs [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+            return await ThumbnailOptions.Where(x => ids.Contains(x.Id)).ToArrayAsync();
+        }
+
+        public async Task RemoveThumbnailTasksByIdsAsync(string[] ids)
+        {
+            var taskEntities = await GetThumbnailTasksByIdsAsync(ids);
+
+            foreach (var taskEntity in taskEntities)
+            {
+                Remove(taskEntity);
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Repositories/ThumbnailRepository.cs]: RemoveAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Repositories/ThumbnailRepository.cs [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+            }
+        }
+
+        public async Task RemoveThumbnailOptionsByIds(string[] ids)
+        {
+            var optionEntities = await GetThumbnailOptionsByIdsAsync(ids);
+
+            foreach (var optionEntity in optionEntities)
+            {
+                Remove(optionEntity);
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Repositories/ThumbnailRepository.cs]: RemoveAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+                            SortColumn = ReflectionUtility.GetPropertyName<ThumbnailTask>(t => t.CreatedDate), SortDirection = SortDirection.Descending
+                        }
+                    };
+
+                var query = repository.ThumbnailOptions.OrderBySortInfos(sortInfos);
+                var totalCount = await query.CountAsync();
+
+                var ids = await query.Skip(criteria.Skip).Take(criteria.Take).Select(x => x.Id).ToArrayAsync();
+                var thumbnailOptions = await repository.GetThumbnailOptionsByIdsAsync(ids);
+                var results = thumbnailOptions.Select(t => t.ToModel(AbstractTypeFactory<ThumbnailOption>.TryCreateInstance())).ToArray();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs]: ToArrayAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskSearchService.cs [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+                            SortColumn = ReflectionUtility.GetPropertyName<ThumbnailTask>(t => t.CreatedDate), SortDirection = SortDirection.Descending
+                        }
+                    };
+                }
+                query = query.OrderBySortInfos(sortInfos);
+                var totalCount = await query.CountAsync();
+
+                var ids = await query.Skip(criteria.Skip).Take(criteria.Take).Select(x => x.Id).ToArrayAsync();
+                var thumbnailTasks = await repository.GetThumbnailTasksByIdsAsync(ids);
+                var results = thumbnailTasks.Select(t => t.ToModel(AbstractTypeFactory<ThumbnailTask>.TryCreateInstance())).ToArray();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskSearchService.cs]: ToArrayAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskService.cs [hunk 31]
```diff
@@ -0,0 +1,69 @@
...
+                pkMap.ResolvePrimaryKeys();
+            }
+        }
+
+        public async Task RemoveByIdsAsync(string[] ids)
+        {
+            using (var repository = _thumbnailRepositoryFactory())
+            {
+                await repository.RemoveThumbnailTasksByIdsAsync(ids);
+                repository.UnitOfWork.Commit();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskService.cs]: CommitAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/ThumbnailGeneration/BlobImagesChangesProvider.cs [hunk 32]
```diff
@@ -0,0 +1,183 @@
...
+        /// Check if image is exist in blob storage by url.
+        /// </summary>
+        /// <param name="imageUrl">Image url.</param>
+        /// <returns>
+        /// EntryState if image exist.
+        /// Null is image is empty
+        /// </returns>
+        protected virtual bool ExistsAsync(string imageUrl)
+        {
+            var blobInfo = _storageProvider.GetBlobInfoAsync(imageUrl).GetAwaiter().GetResult();
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/ThumbnailGeneration/BlobImagesChangesProvider.cs]: GetAwaiter().GetResult() -> await

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/ThumbnailGeneration/DefaultThumbnailGenerator.cs [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+            }
+
+            foreach (var option in options)
+            {
+                var thumbnail = GenerateThumbnail(clone, option);
+                var thumbnailUrl = sourcePath.GenerateThumbnailName(option.FileSuffix);
+
+                if (thumbnail != null)
+                {
+                    await _imageService.SaveImage(thumbnailUrl, thumbnail, format);
```
Comment (tatarincev) [inline Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/ThumbnailGeneration/DefaultThumbnailGenerator.cs]: SaveImageAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs (hunk 29 above)
Comment (n2pro) [reply Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailOptionSearchService.cs]: The method uses the implementation ToArrayAsync. This I uses linq

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskSearchService.cs (hunk 30 above)
Comment (n2pro) [reply Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Services/ThumbnailTaskSearchService.cs]: The method uses the implementation ToArrayAsync. This I uses linq

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Repositories/ThumbnailRepository.cs (hunk 27 above)
Comment (n2pro) [reply Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Repositories/ThumbnailRepository.cs]: In ef core there is no implementation RemoveAsync

File: Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Repositories/ThumbnailRepository.cs (hunk 28 above)
Comment (n2pro) [reply Modules/vc-module-imageTools/VirtoCommerce.ImageToolsModule.Data/Repositories/ThumbnailRepository.cs]: In ef core there is no implementation RemoveAsync

**Pull request merged**